pkg/controllers/leaderworkerset_controller.go

```diff
@@ -523,8 +549,72 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
 	return continuousReadyReplicas, lwsUnreadyReplicas, nil
 }
 
+func (r *LeaderWorkerSetReconciler) getLeaderStatefulSet(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet) (*appsv1.StatefulSet, error) {
+	sts := &appsv1.StatefulSet{}
+	err := r.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, sts)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	return sts, nil
+}
+
+// Creates a Controller Revision if the leader statefulset exists but no revisions have been created yet. This happens when updating from a version that doesn't
+// support controller revision
+func (r *LeaderWorkerSetReconciler) createControllerRevisionIfNonExist(ctx context.Context, sts *appsv1.StatefulSet, lws *leaderworkerset.LeaderWorkerSet) error {
+
+	if sts == nil {
+		return nil
+	}
+
+	existingControllerRevisions, err := revisionutils.ExistingControllerRevisions(ctx, r.Client, lws)
+	if err != nil {
+		return err
+	}
+
+	if !existingControllerRevisions {
+		return revisionutils.CreateLeaderWorkerSetRevision(ctx, r.Client, lws, sts.Labels[leaderworkerset.TemplateRevisionHashKey])
+	}
+
+	return nil
+}
+
+func (r *LeaderWorkerSetReconciler) leaderWorkerSetUpdated(ctx context.Context, sts *appsv1.StatefulSet, lws *leaderworkerset.LeaderWorkerSet) (bool, error) {
+
+	if sts == nil {
+		return false, nil
+	}
+
+	controllerRevision, err := revisionutils.GetLeaderWorkerSetRevisionFromTemplateHash(ctx, r.Client, lws, sts.Labels[leaderworkerset.TemplateRevisionHashKey])
+	if err != nil {
+		return false, err
+	}
+	baselineLws, err := revisionutils.ApplyRevision(lws, controllerRevision)
+	if err != nil {
+		return false, err
+	}
+	return !revisionutils.EqualLeaderWorkerTemplates(baselineLws, lws), nil
```

Comment:
We should use `EqualRevision` here instead. We compare the existing revision (one fetched in line 591) with one we create in memory from the current lws using `NewRevision`.

We don't want another equality function, `EqualLeaderWorkerTemplates` should not exist.